ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -767,16 +768,27 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
               }
             } else if (srcValue.isUndelete()) {
               if (valueFromTgtIdx != null) {
-                tgtIndex.markAsUndeleted(srcIndexEntry.getKey(), fileSpan, srcValue.getOperationTimeInMs(),
-                    srcValue.getLifeVersion());
-              } else {
-                IndexValue tgtValue = new IndexValue(srcValue.getSize(), fileSpan.getStartOffset(), srcValue.getFlags(),
-                    srcValue.getExpiresAtMs(), srcValue.getOperationTimeInMs(), srcValue.getAccountId(),
-                    srcValue.getContainerId(), srcValue.getLifeVersion());
-                tgtValue.setFlag(IndexValue.Flags.Undelete_Index);
-                tgtValue.clearOriginalMessageOffset();
-                tgtIndex.addToIndex(new IndexEntry(srcIndexEntry.getKey(), tgtValue), fileSpan);
+                // Here don't use markAsUndelete method. In markAsUndelete, multiple sanity check would be applied. But
+                // target index might only contain incomplete blob history, which would fail sanity check.
+                // Here we do a different check to make sure we don't insert undelete with lower lifeVersion.
+                if (valueFromTgtIdx.isUndelete()) {
+                  throw new StoreException(
+                      "Id " + srcIndexEntry.getKey() + " already has undelete " + valueFromTgtIdx + " in index "
+                          + dataDir, StoreErrorCodes.ID_Undeleted);
+                }
+                // Undelete would increase the life version
+                if (valueFromTgtIdx.getLifeVersion() >= srcValue.getLifeVersion()) {
+                  throw new StoreException(
+                      "Id " + srcIndexEntry.getKey() + " has bad lifeversion " + valueFromTgtIdx + " in index "
+                          + dataDir, StoreErrorCodes.Life_Version_Conflict);
+                }
```

Comment:
Looks like if `valueFromTgtIdx != null`, we only check version and undelete state, and we still add a new index entry to `tgtIndex`?

Reply:
We are throwing an error when the `valueFromTgtIndex` is not what we expect. The idea is that, in target index, if there are already index value existed for this key, then it can't be an undelete. Compaction would remove all previous undelete in the same segment. And it's lifeVersion can't be equal or greater than the undelete's lifeVersion. It' pretty easy to understand why it can't be greater than undelete, for example if we have PUT in the target index, and it's lifeVersion has to be smaller than the undelete. Also it can't be equal to undelete, because undelete always increase the lifeVersion.